tiledb/sm/misc/resource_pool.h

```diff
@@ -139,6 +144,94 @@ class ResourcePool {
   friend class ResourceHandle<T, tiledb::sm::ResourcePool>;
 };
 
+template <class T>
+class BlockingResourcePool {
+ public:
+  using resource_handle = ResourceHandle<T, tiledb::sm::BlockingResourcePool>;
+
+  /** Constructor. */
+  BlockingResourcePool(unsigned int n)
+      : resources_(n)
+      , unused_(n)
+      , unused_idx_(n - 1)
+      , num_blocked_threads_(0) {
+    for (unsigned int i = 0; i < n; i++)
+      unused_[i] = i;
+  }
+
+  /** Take a resource from the pool. */
+  resource_handle take() {
+    if (resources_exhausted()) {
+      // block until available again
+      std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
+      num_blocked_threads_++;
+      exhaustion_cv_.wait(e_lck, [this]() { return !resources_exhausted(); });
+      num_blocked_threads_--;
+    }
+    return add_resource();
```

Comment:
You'll need to check resource availability once more, but this time with a single lifespan covering both the predicate check and the resource grant.  It's possible for `unused_idx_` to change value between the time that `wait()` returns and when `add_resource` is called.

This is to cover the case when there are more unblocked threads than available resources. This is the only place `add_resource()` is called, so it may be better to eliminate it as a separate function to get the right flow control here.

Reply:
ok